### ISSUE TITLE
Wrong type interference with None

### ISSUE THREAD

**Issue by DerWeh** [PyCQA/pylint#3530]
### Steps to reproduce
Example script `test.py`:
```python
from typing import Optional


def test(opt: Optional[int] = None):
    opt = 1 if opt is None else opt
    return -opt


def test_explicit(opt: Optional[int] = None):
    opt: int = 1 if opt is None else opt
    return -opt
```
### Current behavior
test.py:6:11: E1130: bad operand type for unary -: NoneType (invalid-unary-operand-type)
test.py:11:11: E1130: bad operand type for unary -: NoneType (invalid-unary-operand-type)

### Expected behavior
`pylint` infers that we have a actually a value for `opt` and doesn't give errors.

In *my* opinion, `pylint` should at least take the hint if I manually tell it, that it's the correct type.
Having some superfluous type declarations are in *my* opinion better, then having many comments disabling the errors if the type is wrongfully inferred.

### pylint --version output
```
pylint 2.5.0
astroid 2.4.0
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0]
```

**Comment by PCManticore** [comment]
Thank you for the report.

This is right now expected as `pylint` does not understand control flow guards nor it does understand type annotations to figure out if a variable should be `Optional` or not.

It's on our roadmap to make `pylint` bit smarter so that `if x is None` guards are taken in consideration, but no work on it has been done so far. I'm closing this issue as there are 30+ similar issues related to the same problem, all living under the `topic-control-flow` label, but it's definitely something that we'll support at some point.
